Hyper/Shaders/ObjectShader/AbstractObjectShader.cs

```diff
@@ -23,31 +23,66 @@ protected AbstractObjectShader(float globalScale)
         GlobalScale = globalScale;
     }
 
-    public void SetCurv(float curv) => SetFloat("curv", curv);
+    private void SetCurv(float curv) => SetFloat("curv", curv);
+
+    private void SetView(Matrix4 view) => SetMatrix4("view", view);
 
-    public void SetView(Matrix4 view) => SetMatrix4("view", view);
+    private void SetProjection(Matrix4 projection) => SetMatrix4("projection", projection);
 
-    public void SetProjection(Matrix4 projection) => SetMatrix4("projection", projection);
+    private void SetNumPointLights(int numPointLights) => SetInt("numPointLights", numPointLights);
 
-    public void SetNumLights(int numLights) => SetInt("numLights", numLights);
+    private void SetNumSpotLights(int numSpotLights) => SetInt("numSpotLights", numSpotLights);
 
-    public void SetViewPos(Vector4 viewPos) => SetVector4("viewPos", viewPos);
+    private void SetViewPos(Vector4 viewPos) => SetVector4("viewPos", viewPos);
 
-    public void SetLightColors(Vector3[] lightColors) => SetVector3Array("lightColor", lightColors);
+    private void SetPointLights(PointLight[] lights) => SetStructArray("pointLights", lights);
 
-    public void SetLightPositions(Vector4[] lightPositions) => SetVector4Array("lightPos", lightPositions);
+    private void SetSpotLights(DataTypes.SpotLight[] lights) => SetStructArray("spotLights", lights);
 
-    public virtual void SetUp(Camera camera, List<LightSource> lightSources, int sphere = 0)
+    private void SetShininess(float shininess) => SetFloat("shininess", shininess);
+
+    public virtual void SetUp(Camera camera, List<Lamp> lightSources, List<FlashLight> flashLights, float shininess, int sphere = 0)
     {
         Use();
         SetCurv(camera.Curve);
         SetView(camera.GetViewMatrix());
         SetProjection(camera.GetProjectionMatrix());
-        SetNumLights(lightSources.Count);
+        SetNumPointLights(lightSources.Count);
+        SetNumSpotLights(flashLights.Where(x => x.Active).Count());
+        SetShininess(shininess);
         SetViewPos(GeomPorting.EucToCurved(camera.ViewPosition, camera.Curve));
 
-        SetLightColors(lightSources.Select(x => x.Color).ToArray());
-        SetLightPositions(lightSources.Select(x =>
-            GeomPorting.EucToCurved(GeomPorting.CreateTranslationTarget(x.Position, camera.ReferencePointPosition, camera.Curve, GlobalScale), camera.Curve)).ToArray());
+        SetPointLights(lightSources.Select(x =>
+           new PointLight
+           {
+               Position = GeomPorting.EucToCurved(GeomPorting.CreateTranslationTarget(x.Position, camera.ReferencePointPosition, camera.Curve, GlobalScale), camera.Curve),
```

Comment:
this whole part is also in model shader. Create some seperate class for this stuff. Also could move stuff like SetProjection to that class if it is in both

Reply:
Moved to `ShaderData`